File: apps/ios/Sources/Localization/Localization.swift
import Foundation
import SwiftUI

struct Language: Equatable, Sendable {
    let tag: String
    let nativeName: String
    let direction: String
}

struct LocalizedMessage {
    let path: String
    let arguments: [String: Any]

    init(_ path: String, arguments: [String: Any] = [:]) {
        self.path = path
        self.arguments = arguments
    }
}

private struct LocalizationEnvironmentKey: EnvironmentKey {
    static let defaultValue = Localization.system()
}

extension EnvironmentValues {
    var localization: Localization {
        get { self[LocalizationEnvironmentKey.self] }
        set { self[LocalizationEnvironmentKey.self] = newValue }
    }
}

enum TemplateToken {
    case text(String)
    case placeholder(String)
}

enum CatalogMessage {
    case plain([TemplateToken])
    case plural(argument: String, variants: [String: [TemplateToken]])
}

struct RuntimeCatalog {
    let tag: String
    let englishName: String
    let nativeName: String
    let direction: String
    let aliases: [String]
    let messages: [String: CatalogMessage]
}

private struct LanguageShape {
    let language: String
    let script: String?
    let hasRegion: Bool
}

final class Localization {
    let effectiveLanguage: Language
    let availableLanguages: [Language]

    private let formatLanguageTag: String
    private let messageCatalogs: [RuntimeCatalog]
    private let currentTimeMillis: () -> Double
    private let reportDiagnostic: (String) -> Void
    private var reportedDiagnostics: Set<String> = []

    static func system(
        requestedLanguageTags: [String] = Locale.preferredLanguages,
        regionalLanguageTag: String = Locale.current.identifier,
        regionalNumberingSystem: Locale.NumberingSystem = Locale.current.numberingSystem
    ) -> Localization {
        return Localization(
            runtimeCatalogs: GeneratedLanguageCatalogs.catalogs,
            requestedLanguageTags: requestedLanguageTags,
            regionalLanguageTag: regionalLanguageTag,
            regionalNumberingSystem: regionalNumberingSystem
        )
    }

    init(
        runtimeCatalogs: [RuntimeCatalog],
        requestedLanguageTags: [String],
        regionalLanguageTag: String?,
        regionalNumberingSystem: Locale.NumberingSystem? = nil,
        currentTimeMillis: @escaping () -> Double = { Date().timeIntervalSince1970 * 1_000 },
        reportDiagnostic: @escaping (String) -> Void = { NSLog("%@", $0) }
    ) {
        let selectedCatalog = Self.selectCatalog(
            requestedLanguageTags: requestedLanguageTags,
            catalogs: runtimeCatalogs
        )
        let effectiveLanguage =
            selectedCatalog.map {
                Language(tag: $0.tag, nativeName: $0.nativeName, direction: $0.direction)
            } ?? Language(tag: "en", nativeName: "English", direction: "ltr")
        let formatLanguageTag = Self.formattingLanguageTag(
            selectedLanguageTag: effectiveLanguage.tag,
            regionalLanguageTag: regionalLanguageTag,
            regionalNumberingSystem: regionalNumberingSystem
        )
        let availableLanguages =
            runtimeCatalogs
            .sorted {
                let englishNameOrder = $0.englishName.compare(
                    $1.englishName,
                    options: [],
                    range: nil,
                    locale: Locale(identifier: "en")
                )
                return englishNameOrder == .orderedSame
                    ? $0.tag < $1.tag
                    : englishNameOrder == .orderedAscending
            }
            .map { Language(tag: $0.tag, nativeName: $0.nativeName, direction: $0.direction) }
        self.effectiveLanguage = effectiveLanguage
        self.availableLanguages = availableLanguages
        self.formatLanguageTag = formatLanguageTag
        self.messageCatalogs = Self.fallbackCatalogs(
            selectedCatalog: selectedCatalog,
            catalogs: runtimeCatalogs
        )
        self.currentTimeMillis = currentTimeMillis
        self.reportDiagnostic = reportDiagnostic
    }

    func localizedText(_ path: String, arguments: [String: Any] = [:]) -> String {
        for catalog in messageCatalogs {
            guard let message = catalog.messages[path] else { continue }
            switch message {
            case .plain(let template):
                return renderTemplate(template, arguments: arguments, path: path)
            case .plural(let pluralArgument, let variants):
                guard let pluralValue = integerArgument(arguments[pluralArgument]), pluralValue >= 0
                else {
                    reportOnce(
                        key: "\(effectiveLanguage.tag):\(path):invalid-plural-argument",
                        message:
                            "Localization catalog error: language=\(effectiveLanguage.tag) path=\(path) type=invalid-plural-argument"
                    )
                    return path
                }
                let template =
                    variants["=\(pluralValue)"]
                    ?? variants[
                        localizationPluralCategory(languageTag: catalog.tag, value: pluralValue)]
                    ?? variants["other"]
                guard let template else { return path }
                return renderTemplate(template, arguments: arguments, path: path)
            }
        }
        reportOnce(
            key: "\(effectiveLanguage.tag):\(path):missing-message",
            message:
                "Localization catalog error: language=\(effectiveLanguage.tag) path=\(path) type=missing-message"
        )
        return path
    }

    func localizedFileSize(_ bytes: Int64) -> String {
        if bytes >= 1_000_000_000_000 {
            let value = (Double(bytes) / 1_000_000_000_000 * 10).rounded() / 10
            return localizedText("units.fileSize.terabyte", arguments: ["value": value])
        }
        if bytes >= 1_000_000_000 {
            let value = (Double(bytes) / 1_000_000_000 * 10).rounded() / 10
            return localizedText("units.fileSize.gigabyte", arguments: ["value": value])
        }
        if bytes >= 1_000_000 {
            let value = (Double(bytes) / 1_000_000 * 10).rounded() / 10
            return localizedText("units.fileSize.megabyte", arguments: ["value": value])
        }
        if bytes >= 1_000 {
            let value = (Double(bytes) / 1_000 * 10).rounded() / 10
            return localizedText("units.fileSize.kilobyte", arguments: ["value": value])
        }
        return localizedText("units.fileSize.byte", arguments: ["value": bytes])
    }

    func localizedRelativeTime(_ timestampMillis: Double) -> String {
        let differenceSeconds = (timestampMillis - currentTimeMillis()) / 1_000
        let absoluteSeconds = abs(differenceSeconds)
        if absoluteSeconds < 60 { return localizedText("time.relative.now") }
        if absoluteSeconds >= 365 * 24 * 60 * 60 {
            let count = UInt64(floor(absoluteSeconds / (365 * 24 * 60 * 60)))
            return differenceSeconds < 0
                ? localizedText("time.relative.past.year", arguments: ["count": count])
                : localizedText("time.relative.future.year", arguments: ["count": count])
        }
        if absoluteSeconds >= 30 * 24 * 60 * 60 {
            let count = UInt64(floor(absoluteSeconds / (30 * 24 * 60 * 60)))
            return differenceSeconds < 0
                ? localizedText("time.relative.past.month", arguments: ["count": count])
                : localizedText("time.relative.future.month", arguments: ["count": count])
        }
        if absoluteSeconds >= 24 * 60 * 60 {
            let count = UInt64(floor(absoluteSeconds / (24 * 60 * 60)))
            return differenceSeconds < 0
                ? localizedText("time.relative.past.day", arguments: ["count": count])
                : localizedText("time.relative.future.day", arguments: ["count": count])
        }
        if absoluteSeconds >= 60 * 60 {
            let count = UInt64(floor(absoluteSeconds / (60 * 60)))
            return differenceSeconds < 0
                ? localizedText("time.relative.past.hour", arguments: ["count": count])
                : localizedText("time.relative.future.hour", arguments: ["count": count])
        }
        let count = UInt64(floor(absoluteSeconds / 60))
        return differenceSeconds < 0
            ? localizedText("time.relative.past.minute", arguments: ["count": count])
            : localizedText("time.relative.future.minute", arguments: ["count": count])
    }

    private func renderTemplate(
        _ template: [TemplateToken],
        arguments: [String: Any],
        path: String
    ) -> String {
        let numberFormatter = NumberFormatter()
        numberFormatter.numberStyle = .decimal
        numberFormatter.locale = Locale(identifier: formatLanguageTag)
        numberFormatter.maximumFractionDigits = 3
        return template.map { token in
            switch token {
            case .text(let value):
                return value
            case .placeholder(let name):
                guard let value = arguments[name] else {
                    reportOnce(
                        key: "\(effectiveLanguage.tag):\(path):missing-argument:\(name)",
                        message:
                            "Localization catalog error: language=\(effectiveLanguage.tag) path=\(path) type=missing-argument name=\(name)"
                    )
                    return "{\(name)}"
                }
                if let string = value as? String { return string }
                if let number = value as? NSNumber {
                    return numberFormatter.string(from: number) ?? number.stringValue
                }
                return String(describing: value)
            }
        }.joined()
    }

    private func integerArgument(_ value: Any?) -> UInt64? {
        guard let number = value as? NSNumber else { return nil }
        let doubleValue = number.doubleValue
        guard doubleValue.isFinite, doubleValue >= 0,
            doubleValue.rounded(.towardZero) == doubleValue
        else {
            return nil
        }
        return UInt64(exactly: doubleValue)
    }

    private func reportOnce(key: String, message: String) {
        if reportedDiagnostics.insert(key).inserted { reportDiagnostic(message) }
    }

    private static func canonicalLanguageTag(_ languageTag: String) -> String? {
        let canonical = Locale.canonicalLanguageIdentifier(from: languageTag)
            .replacingOccurrences(of: "_", with: "-")
        guard !canonical.isEmpty, localizationMaximizeLanguageTag(languageTag: canonical) != nil
        else {
            return nil
        }
        return canonical
    }

    private static func languageShape(_ languageTag: String, maximize: Bool) -> LanguageShape? {
        let resolvedTag: String
        if maximize {
            guard let maximized = localizationMaximizeLanguageTag(languageTag: languageTag) else {
                return nil
            }
            resolvedTag = maximized
        } else {
            resolvedTag = languageTag
        }
        let components = Locale.Language.Components(identifier: resolvedTag)
        guard let language = components.languageCode?.identifier else { return nil }
        return LanguageShape(
            language: language,
            script: components.script?.identifier,
            hasRegion: components.region != nil
        )
    }

    private static func compatibleCatalogs(
        requestedLanguageTag: String,
        catalogs: [RuntimeCatalog]
    ) -> [RuntimeCatalog] {
        guard let requested = languageShape(requestedLanguageTag, maximize: true),
            let requestedScript = requested.script
        else { return [] }
        return catalogs.filter { catalog in
            guard let candidate = languageShape(catalog.tag, maximize: true) else { return false }
            return candidate.language == requested.language && candidate.script == requestedScript
        }
    }

    private static func selectRequestedCatalog(
        _ requestedLanguageTag: String,
        catalogs: [RuntimeCatalog]
    ) -> RuntimeCatalog? {
        guard let canonicalTag = canonicalLanguageTag(requestedLanguageTag) else { return nil }
        if let exact = catalogs.first(where: { $0.tag == canonicalTag }) { return exact }
        if let alias = catalogs.first(where: { $0.aliases.contains(canonicalTag) }) { return alias }
        let compatible = compatibleCatalogs(requestedLanguageTag: canonicalTag, catalogs: catalogs)
        let generic = compatible.filter {
            languageShape($0.tag, maximize: false)?.hasRegion == false
        }
        if generic.count == 1 { return generic[0] }
        return compatible.count == 1 ? compatible[0] : nil
    }

    private static func selectCatalog(
        requestedLanguageTags: [String],
        catalogs: [RuntimeCatalog]
    ) -> RuntimeCatalog? {
        for requestedLanguageTag in requestedLanguageTags {
            if let catalog = selectRequestedCatalog(requestedLanguageTag, catalogs: catalogs) {
                return catalog
            }
        }
        return catalogs.first(where: { $0.tag == "en" })
    }

    private static func fallbackCatalogs(
        selectedCatalog: RuntimeCatalog?,
        catalogs: [RuntimeCatalog]
    ) -> [RuntimeCatalog] {
        var candidates: [RuntimeCatalog] = []
        if let selectedCatalog {
            candidates.append(selectedCatalog)
            let generic = compatibleCatalogs(
                requestedLanguageTag: selectedCatalog.tag,
                catalogs: catalogs
            ).filter {
                $0.tag != selectedCatalog.tag
                    && languageShape($0.tag, maximize: false)?.hasRegion == false
            }
            if generic.count == 1 { candidates.append(generic[0]) }
        }
        if let english = catalogs.first(where: { $0.tag == "en" }) { candidates.append(english) }
        var seen: Set<String> = []
        return candidates.filter { seen.insert($0.tag).inserted }
    }

    private static func formattingLanguageTag(
        selectedLanguageTag: String,
        regionalLanguageTag: String?,
        regionalNumberingSystem: Locale.NumberingSystem?
    ) -> String {
        var selected = Locale.Components(identifier: selectedLanguageTag)
        let regionalComponents = regionalLanguageTag.map(Locale.Components.init(identifier:))
        selected.region = regionalComponents?.region ?? selected.region
        if let numberingSystem = regionalNumberingSystem
            ?? regionalLanguageTag.map({ Locale(identifier: $0).numberingSystem })
        {
            selected.numberingSystem = numberingSystem
        }
        return Locale(components: selected).identifier
    }
}

File: apps/android/app/src/main/java/com/futo/notes/localization/Localization.kt
package com.futo.notes.localization

import android.content.res.Resources
import android.icu.text.NumberingSystem
import android.icu.util.ULocale

data class Language(
    val tag: String,
    val nativeName: String,
    val direction: String,
)

internal data class LanguageShape(
    val language: String,
    val script: String?,
    val hasRegion: Boolean,
)

internal sealed interface TemplateToken {
    data class Text(val value: String) : TemplateToken
    data class Placeholder(val name: String) : TemplateToken
}

internal sealed interface CatalogMessage {
    data class Plain(val template: List<TemplateToken>) : CatalogMessage
    data class Plural(
        val pluralArgument: String,
        val variants: Map<String, List<TemplateToken>>,
    ) : CatalogMessage
}

internal data class RuntimeCatalog(
    val tag: String,
    val englishName: String,
    val nativeName: String,
    val direction: String,
    val aliases: List<String>,
    val messages: Map<String, CatalogMessage>,
)

class Localization private constructor(
    private val catalogs: List<RuntimeCatalog>,
    requestedLanguageTags: List<String>,
    regionalLanguageTag: String?,
    regionalNumberingSystem: String?,
    private val currentTimeMillis: () -> Long,
    private val reportDiagnostic: (String) -> Unit,
) {
    private val reportedDiagnostics = mutableSetOf<String>()
    private val selectedCatalog = selectCatalog(requestedLanguageTags)
    private val formatLanguageTag = AndroidLocalizationRules.formattingLanguageTag(
        selectedCatalog?.tag ?: "en",
        regionalLanguageTag,
        regionalNumberingSystem,
    )
    private val messageCatalogs = fallbackCatalogs()

    val effectiveLanguage = selectedCatalog?.let {
        Language(it.tag, it.nativeName, it.direction)
    } ?: Language("en", "English", "ltr")

    val availableLanguages = catalogs
        .sortedWith { left, right ->
            val englishNameOrder = AndroidLocalizationRules.compare("en", left.englishName, right.englishName)
            englishNameOrder.takeIf { it != 0 } ?: left.tag.compareTo(right.tag)
        }
        .map { Language(it.tag, it.nativeName, it.direction) }

    fun localizedText(path: String, arguments: Map<String, Any> = emptyMap()): String {
        for (catalog in messageCatalogs) {
            when (val message = catalog.messages[path]) {
                is CatalogMessage.Plain -> {
                    return renderTemplate(message.template, arguments, path)
                }
                is CatalogMessage.Plural -> {
                    val pluralValue = integerArgument(arguments[message.pluralArgument])
                    if (pluralValue == null || pluralValue < 0) {
                        reportOnce(
                            "${effectiveLanguage.tag}:$path:invalid-plural-argument",
                            "Localization catalog error: language=${effectiveLanguage.tag} path=$path type=invalid-plural-argument",
                        )
                        return path
                    }
                    val exactSelector = "=$pluralValue"
                    val template = message.variants[exactSelector]
                        ?: message.variants[AndroidLocalizationRules.pluralCategory(catalog.tag, pluralValue)]
                        ?: message.variants.getValue("other")
                    return renderTemplate(template, arguments, path)
                }
                null -> continue
            }
        }
        reportOnce(
            "${effectiveLanguage.tag}:$path:missing-message",
            "Localization catalog error: language=${effectiveLanguage.tag} path=$path type=missing-message",
        )
        return path
    }

    fun localizedFileSize(bytes: Long): String {
        if (bytes >= 1_000_000_000_000L) {
            val value = kotlin.math.floor(bytes.toDouble() / 1_000_000_000_000L * 10.0 + 0.5) / 10.0
            return localizedText("units.fileSize.terabyte", mapOf("value" to value))
        }
        if (bytes >= 1_000_000_000L) {
            val value = kotlin.math.floor(bytes.toDouble() / 1_000_000_000L * 10.0 + 0.5) / 10.0
            return localizedText("units.fileSize.gigabyte", mapOf("value" to value))
        }
        if (bytes >= 1_000_000L) {
            val value = kotlin.math.floor(bytes.toDouble() / 1_000_000L * 10.0 + 0.5) / 10.0
            return localizedText("units.fileSize.megabyte", mapOf("value" to value))
        }
        if (bytes >= 1_000L) {
            val value = kotlin.math.floor(bytes.toDouble() / 1_000L * 10.0 + 0.5) / 10.0
            return localizedText("units.fileSize.kilobyte", mapOf("value" to value))
        }
        return localizedText("units.fileSize.byte", mapOf("value" to bytes))
    }

    fun localizedRelativeTime(timestampMillis: Long): String {
        val differenceSeconds = (timestampMillis - currentTimeMillis()) / 1_000.0
        val absoluteSeconds = kotlin.math.abs(differenceSeconds)
        if (absoluteSeconds < 60) return localizedText("time.relative.now")
        if (absoluteSeconds >= 365L * 24 * 60 * 60) {
            val count = kotlin.math.floor(absoluteSeconds / (365L * 24 * 60 * 60)).toLong()
            return if (differenceSeconds < 0) {
                localizedText("time.relative.past.year", mapOf("count" to count))
            } else {
                localizedText("time.relative.future.year", mapOf("count" to count))
            }
        }
        if (absoluteSeconds >= 30L * 24 * 60 * 60) {
            val count = kotlin.math.floor(absoluteSeconds / (30L * 24 * 60 * 60)).toLong()
            return if (differenceSeconds < 0) {
                localizedText("time.relative.past.month", mapOf("count" to count))
            } else {
                localizedText("time.relative.future.month", mapOf("count" to count))
            }
        }
        if (absoluteSeconds >= 24L * 60 * 60) {
            val count = kotlin.math.floor(absoluteSeconds / (24L * 60 * 60)).toLong()
            return if (differenceSeconds < 0) {
                localizedText("time.relative.past.day", mapOf("count" to count))
            } else {
                localizedText("time.relative.future.day", mapOf("count" to count))
            }
        }
        if (absoluteSeconds >= 60L * 60) {
            val count = kotlin.math.floor(absoluteSeconds / (60L * 60)).toLong()
            return if (differenceSeconds < 0) {
                localizedText("time.relative.past.hour", mapOf("count" to count))
            } else {
                localizedText("time.relative.future.hour", mapOf("count" to count))
            }
        }
        val count = kotlin.math.floor(absoluteSeconds / 60L).toLong()
        return if (differenceSeconds < 0) {
            localizedText("time.relative.past.minute", mapOf("count" to count))
        } else {
            localizedText("time.relative.future.minute", mapOf("count" to count))
        }
    }

    private fun renderTemplate(
        template: List<TemplateToken>,
        arguments: Map<String, Any>,
        path: String,
    ): String = buildString {
        for (token in template) {
            when (token) {
                is TemplateToken.Text -> append(token.value)
                is TemplateToken.Placeholder -> {
                    val value = arguments[token.name]
                    if (value == null) {
                        reportOnce(
                            "${effectiveLanguage.tag}:$path:missing-argument:${token.name}",
                            "Localization catalog error: language=${effectiveLanguage.tag} path=$path type=missing-argument name=${token.name}",
                        )
                        append("{${token.name}}")
                    } else if (value is Number) {
                        append(AndroidLocalizationRules.formatNumber(formatLanguageTag, value))
                    } else {
                        append(value.toString())
                    }
                }
            }
        }
    }

    private fun integerArgument(value: Any?): Long? {
        if (value !is Number) return null
        val doubleValue = value.toDouble()
        if (!doubleValue.isFinite() || doubleValue % 1.0 != 0.0) return null
        return doubleValue.toLong()
    }

    private fun reportOnce(key: String, message: String) {
        if (reportedDiagnostics.add(key)) reportDiagnostic(message)
    }

    private fun compatibleCatalogs(requestedLanguageTag: String): List<RuntimeCatalog> {
        val requested = AndroidLocalizationRules.languageShape(requestedLanguageTag, maximize = true)
        if (requested?.script == null) return emptyList()
        return catalogs.filter { catalog ->
            val candidate = AndroidLocalizationRules.languageShape(catalog.tag, maximize = true)
            candidate?.language == requested.language && candidate.script == requested.script
        }
    }

    private fun selectRequestedCatalog(requestedLanguageTag: String): RuntimeCatalog? {
        val canonicalTag = AndroidLocalizationRules.canonicalLanguageTag(requestedLanguageTag) ?: return null
        catalogs.firstOrNull { it.tag == canonicalTag }?.let { return it }
        catalogs.firstOrNull { canonicalTag in it.aliases }?.let { return it }
        val compatible = compatibleCatalogs(canonicalTag)
        val generic = compatible.filter {
            AndroidLocalizationRules.languageShape(it.tag, maximize = false)?.hasRegion == false
        }
        if (generic.size == 1) return generic.first()
        return compatible.singleOrNull()
    }

    private fun selectCatalog(requestedLanguageTags: List<String>): RuntimeCatalog? {
        for (requestedLanguageTag in requestedLanguageTags) {
            selectRequestedCatalog(requestedLanguageTag)?.let { return it }
        }
        return catalogs.firstOrNull { it.tag == "en" }
    }

    private fun fallbackCatalogs(): List<RuntimeCatalog> {
        val candidates = mutableListOf<RuntimeCatalog>()
        selectedCatalog?.let(candidates::add)
        selectedCatalog?.let { selected ->
            val generic = compatibleCatalogs(selected.tag).filter {
                it.tag != selected.tag && AndroidLocalizationRules.languageShape(it.tag, maximize = false)?.hasRegion == false
            }
            generic.singleOrNull()?.let(candidates::add)
        }
        catalogs.firstOrNull { it.tag == "en" }?.let(candidates::add)
        return candidates.distinctBy { it.tag }
    }

    companion object {
        fun system(
            requestedLanguageTags: List<String>,
            regionalLanguageTag: String = Resources.getSystem().configuration.locales[0]
                .toLanguageTag(),
            regionalNumberingSystem: String = NumberingSystem.getInstance(
                ULocale.forLanguageTag(regionalLanguageTag),
            ).name,
            reportDiagnostic: (String) -> Unit = { message ->
                android.util.Log.e("FutoNotesLocalization", message)
            },
        ): Localization = fromGeneratedCatalogs(
            requestedLanguageTags,
            regionalLanguageTag,
            regionalNumberingSystem,
            reportDiagnostic = reportDiagnostic,
        )

        internal fun fromRuntimeCatalogs(
            catalogs: List<RuntimeCatalog>,
            requestedLanguageTags: List<String>,
            regionalLanguageTag: String?,
            regionalNumberingSystem: String? = null,
            currentTimeMillis: () -> Long = System::currentTimeMillis,
            reportDiagnostic: (String) -> Unit = {},
        ): Localization = Localization(
            catalogs,
            requestedLanguageTags,
            regionalLanguageTag,
            regionalNumberingSystem,
            currentTimeMillis,
            reportDiagnostic,
        )

        internal fun fromGeneratedCatalogs(
            requestedLanguageTags: List<String>,
            regionalLanguageTag: String?,
            regionalNumberingSystem: String? = null,
            currentTimeMillis: () -> Long = System::currentTimeMillis,
            reportDiagnostic: (String) -> Unit = {},
        ): Localization {
            return Localization(
                GeneratedLanguageCatalogs.catalogs,
                requestedLanguageTags,
                regionalLanguageTag,
                regionalNumberingSystem,
                currentTimeMillis,
                reportDiagnostic,
            )
        }
    }
}
